docs/src/index.md

```diff
@@ -6,4 +6,369 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+extern(ċ)
+# -2.0638950738662625
+
+###
+# Find dfoo/dx via finite-difference
+using FiniteDifferences
+central_fdm(5,1)(foo, x)
+# -2.0638950738670734
+
+###
+# Via ForwardDiff.jl
+using ForwardDiff
+ForwardDiff.derivative(foo, x)
+# -2.0638950738662625
+
+###
+# Via Zygote
+using Zygote
+Zygote.gradient(foo, x)
+# (-2.0638950738662625,)
+```
+
+
+ -------------------------------
+
+
+
+## On writing good rrule / frules
+
+### Return Zero or One
+rather than `0` or `1`
+or even rather than `zeros(n)`, `ones(m,n)`
+
+### Use thunks appropriately:
+
+If work is only required for one of the returned differentials it should be wrapped in a `@thunk` (potentially using a begin-end block)
+
+If there are multiple return values, almost always their should be computation wrapped in a `@thunk`s
+
+Don’t wrap variables in thunks, wrap the computations that fill those variables in thunks: Eg:
+Write:
+```julia
+∂A = @thunk(foo(x))
+return ∂A
+```
+Not:
+```julia
+∂A = foo(x)
+return @thunk(∂A)
+```
+In the bad example `foo(x)` gets computed eagerly, and all that the thunk is doing is wrapping the already calculated result in a function that returns it.
+
+### Be careful with using Adjoint when you mean Transpose
+
+Rember for complex numbers `a'` (i.e. `adjoint(a)`) takes the complex conjugate. Instead you probably want `transpose(a)`.
+
+While there are arguments that for reverse-mode
+taking the adjoint is correct, it is not currently the behavour of ChainRules to do so.
+Feel free to open an issue and fight about it.
+All differentials support `conj` efficiently, which makes it easy to change in post.
+
+### Style
+
+Used named local functions for the pushforward/pullback:
+Rather than:
+```julia
+function frule(::typeof(foo), x)
+        return (foo(x), (_, ẋ)->bar(ẋ))
+end
+```
+Whichrite:
+```julia
+function frule(::typeof(foo), x)
+        Y = foo(x)
+        function foo_pushforward(_, ẋ)
+            return bar(ẋ)
+        end
+        return Y, foo_pushforward
+end
+```
+
+
+While this is more verbose,
+it ensures that if an error is thrown during the pullback/pushforward
+the gensymed name of the local function will include the name you gave it.
+Which makes it a lot simpler to debug from the stacktrace.
+
+### Write tests
+There are faily decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl)
+They are in [`tests/test_utils.jl`](https://github.com/JuliaDiff/ChainRules.jl/blob/master/test/test_util.jl)
+Take a look at existing test and you should see how to do stuff.
+
+!!! important
+    Don't write equations in tests.
+    Use finite differencing.
+    If you are writing equations in the tests, then you use those same equations as use are using to write your code. Then that is not Ok. We've had several bugs from people misreading/misunderstanding equations, and then using them for both tests and code. And then we have good coverage that is worthless.
```

Comment:
```
    If you are writing equations in the tests, then you use those same equations as use are using to write your code. Then that is not Ok. 
```

Not sure what this is trying to say really.

Reply:
The problem was 
prior to https://github.com/JuliaDiff/ChainRules.jl/issues/87
we actually had tests for 
`acshc`, `asec`, `acsc`, `asecd`, and  `acscd` 
but they were wrong in the same way that the definitions were wrong.

Reply:
"Since the tests are often written by the same person who implemented the rule it's likely that a mistake in the analytical expression used in the implementation will also be present in the test, thus rendering the test useless." or something?